File: src/main/java/com/dude/dms/backend/data/base/Word.java
package com.dude.dms.backend.data.base;

import com.dude.dms.backend.data.DataEntity;
import com.fasterxml.jackson.annotation.JsonIdentityInfo;
import com.fasterxml.jackson.annotation.ObjectIdGenerators;

import javax.persistence.Entity;
import javax.persistence.ManyToOne;
import javax.validation.constraints.NotNull;

@JsonIdentityInfo(generator = ObjectIdGenerators.PropertyGenerator.class, property = "id")
@Entity
public class Word extends DataEntity {

    @NotNull
    @ManyToOne
    private Doc doc;

    @NotNull
    protected String text;

    @NotNull
    protected Float x;

    @NotNull
    protected Float y;

    @NotNull
    protected Float width;

    @NotNull
    protected Float height;

    public Word() {

    }

    public Word(@NotNull String text, @NotNull Float x, @NotNull Float y, @NotNull Float width, @NotNull Float height) {
        this.text = text;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    public Word(@NotNull Doc doc, @NotNull String text, @NotNull Float x, @NotNull Float y, @NotNull Float width, @NotNull Float height) {
        this.doc = doc;
        this.text = text;
        this.x = x;
        this.y = y;
        this.width = width;
        this.height = height;
    }

    public Doc getDoc() {
        return doc;
    }

    public void setDoc(Doc doc) {
        this.doc = doc;
    }

    public String getText() {
        return text;
    }

    public void setText(String text) {
        this.text = text;
    }

    public Float getX() {
        return x;
    }

    public void setX(Float x) {
        this.x = x;
    }

    public Float getY() {
        return y;
    }

    public void setY(Float y) {
        this.y = y;
    }

    public Float getWidth() {
        return width;
    }

    public void setWidth(Float width) {
        this.width = width;
    }

    public Float getHeight() {
        return height;
    }

    public void setHeight(Float height) {
        this.height = height;
    }

    @Override
    public String toString() {
        return "Word{text='" + text + "', x=" + x + ", y=" + y + ", width=" + width + ", height=" + height + '}';
    }
}

File: src/main/java/com/dude/dms/backend/data/docs/Word.kt
package com.dude.dms.backend.data.docs

import com.dude.dms.backend.data.DataEntity
import com.dude.dms.backend.data.LogsEvents
import com.dude.dms.brain.t
import com.fasterxml.jackson.annotation.JsonIdentityInfo
import com.fasterxml.jackson.annotation.ObjectIdGenerators.PropertyGenerator
import javax.persistence.Column
import javax.persistence.Entity
import javax.persistence.ManyToOne
import javax.validation.constraints.Size

@JsonIdentityInfo(generator = PropertyGenerator::class, property = "id")
@Entity
class Word(
        @ManyToOne var line: Line?,

        @Column(columnDefinition = "LONGVARCHAR")
        @Size(max = 1000)
        var text: String,

        var x: Float,
        var y: Float,
        var width: Float,
        var height: Float
) : DataEntity(), LogsEvents {

    override fun showEvents() = false

    override fun toString() = t("word")

    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (javaClass != other?.javaClass) return false

        other as Word

        if (id > 0 && other.id > 0) return super.equals(other)
        if (text != other.text) return false
        if (x != other.x) return false
        if (y != other.y) return false
        if (width != other.width) return false
        if (height != other.height) return false

        return true
    }

    override fun hashCode(): Int {
        var result = super.hashCode()
        result = 31 * result + text.hashCode()
        result = 31 * result + x.hashCode()
        result = 31 * result + y.hashCode()
        result = 31 * result + width.hashCode()
        result = 31 * result + height.hashCode()
        return result
    }
}